expression/builtin_json.go

```diff
@@ -685,8 +686,51 @@ type jsonValidFunctionClass struct {
 	baseFunctionClass
 }
 
+type builtinJSONValidSig struct {
+	baseBuiltinFunc
+}
+
+func (b *builtinJSONValidSig) Clone() builtinFunc {
+	newSig := &builtinJSONValidSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
 func (c *jsonValidFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
-	return nil, errFunctionNotExists.GenWithStackByArgs("FUNCTION", "JSON_VALID")
+	if err := c.verifyArgs(args); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	bf := newBaseBuiltinFuncWithTp(ctx, args, types.ETInt, args[0].GetType().EvalType())
+	sig := &builtinJSONValidSig{bf}
+	sig.setPbCode(tipb.ScalarFuncSig_JsonValidSig)
+	return sig, nil
+}
+
+func (b *builtinJSONValidSig) evalInt(row chunk.Row) (res int64, isNull bool, err error) {
+	switch b.args[0].GetType().Tp {
```

Comment:
these `switch ... case ...` should be separated into different function signatures. for example:
- use `builtinJSONValidJSONSig` to cover the JSON argument
- use `builtinJSONValidStringSig` to cover the string argument

by doing this, we can reduce the CPU branch predicates and improve the execution performance.

Reply:
> these `switch ... case ...` should be separated into different function signatures. for example:
> 
> * use `builtinJSONValidJSONSig` to cover the JSON argument
> * use `builtinJSONValidStringSig` to cover the string argument
> 
> by doing this, we can reduce the CPU branch predicates and improve the execution performance.

Thank you for your guidance.  I had refer math [ceil function](https://github.com/pingcap/tidb/blob/master/expression/builtin_math.go#L77). And i want to add four function signatures:

- `builtinJSONValidJSONSig` to cover the JSON argument
- `builtinJSONValidStringSig` to cover the string argument
- `builtinJSONValidNullSig` to cover the NULL argument
- `builtinJSONValidOthersSig` to cover the others

Is that OK? And if that is ok, should i remove [JsonValidSig](https://github.com/pingcap/tipb/blob/master/proto/expression.proto#L405) and add related expression proto code   
in tipb?

Reply:
I think we only need two signatures, one for the string argument, the other for the Json argument. And handle null input in these signatures. 

Yes, we should remove the JsonValidSig and add the related proto code in the tipb.

Reply:
Got it.